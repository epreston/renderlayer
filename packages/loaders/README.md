# @renderlayer/loaders

[![NPM version][npm-badge]][npm-url]

[npm-badge]: https://img.shields.io/npm/v/@renderlayer/loaders
[npm-url]: https://www.npmjs.com/package/@renderlayer/loaders
