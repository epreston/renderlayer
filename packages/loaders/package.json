{
  "name": "@renderlayer/loaders",
  "version": "0.0.0",
  "description": "@renderlayer/loaders",
  "main": "index.js",
  "files": [
    "index.js",
    "dist"
  ],
  "sideEffects": false,
  "buildOptions": {
    "formats": [
      "esm-bundler"
    ]
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/epreston/renderlayer.git",
    "directory": "packages/loaders"
  },
  "keywords": [
    "renderlayer"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/epreston/renderlayer/issues"
  },
  "homepage": "https://github.com/epreston/renderlayer/blob/main/packages/loaders#readme"
}
