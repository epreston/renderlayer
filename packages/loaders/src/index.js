export * from './Loader.js';
export * from './LoadingManager.js';
// export * from './ObjectLoader.js'; // execlude from build
