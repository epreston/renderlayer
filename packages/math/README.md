# @renderlayer/math

Math library used by `@renderlayer` packages.

[![NPM version][npm-badge]][npm-url]

[npm-badge]: https://img.shields.io/npm/v/@renderlayer/math
[npm-url]: https://www.npmjs.com/package/@renderlayer/math
