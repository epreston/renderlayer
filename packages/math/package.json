{
  "type": "module",
  "sideEffects": false,
  "name": "@renderlayer/math",
  "version": "0.0.4",
  "description": "@renderlayer/math",
  "module": "./dist/math.esm-bundler.js",
  "exports": "./dist/math.esm-bundler.js",
  "files": [
    "dist"
  ],
  "buildOptions": {
    "formats": [
      "esm-bundler"
    ]
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/epreston/renderlayer.git",
    "directory": "packages/math"
  },
  "keywords": [
    "renderlayer", "math"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/epreston/renderlayer/issues"
  },
  "homepage": "https://github.com/epreston/renderlayer/blob/main/packages/math#readme"
}
