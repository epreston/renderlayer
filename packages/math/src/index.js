export * from './Box2.js';
export * from './Box3.js';
export * from './Color.js';
export * from './ColorManagement.js';
export * from './Euler.js';
export * from './Frustum.js';
export * from './Line3.js';
export * from './MathUtils.js';
export * from './Matrix3.js';
export * from './Matrix4.js';
export * from './Plane.js';
export * from './Quaternion.js';
export * from './Ray.js';
export * from './Sphere.js';
export * from './Triangle.js';
export * from './Vector2.js';
export * from './Vector3.js';
export * from './Vector4.js';
