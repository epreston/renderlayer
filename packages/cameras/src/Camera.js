import { WebGLCoordinateSystem } from '@renderlayer/shared';
import { Matrix4 } from '@renderlayer/math';
import { Object3D } from '@renderlayer/core';

class Camera extends Object3D {
  constructor() {
    super();

    this.isCamera = true;
    this.type = 'Camera';

    this.matrixWorldInverse = new Matrix4();

    this.projectionMatrix = new Matrix4();
    this.projectionMatrixInverse = new Matrix4();

    this.coordinateSystem = WebGLCoordinateSystem;
  }

  copy(source, recursive) {
    super.copy(source, recursive);

    this.matrixWorldInverse.copy(source.matrixWorldInverse);

    this.projectionMatrix.copy(source.projectionMatrix);
    this.projectionMatrixInverse.copy(source.projectionMatrixInverse);

    this.coordinateSystem = source.coordinateSystem;

    return this;
  }

  getWorldDirection(target) {
    return super.getWorldDirection(target).negate();
  }

  updateMatrixWorld(force) {
    super.updateMatrixWorld(force);

    this.matrixWorldInverse.copy(this.matrixWorld).invert();
  }

  updateWorldMatrix(updateParents, updateChildren) {
    super.updateWorldMatrix(updateParents, updateChildren);

    this.matrixWorldInverse.copy(this.matrixWorld).invert();
  }

  /** @returns {this} */
  clone() {
    return new this.constructor().copy(this);
  }
}

export { Camera };
