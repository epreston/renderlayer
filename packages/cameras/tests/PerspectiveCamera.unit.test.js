import { describe, expect, it, test, vi } from 'vitest';

import { ObjectLoader } from '@renderlayer/loaders';
import { Matrix4 } from '@renderlayer/math';

import { Camera } from '../src/Camera.js';
import { PerspectiveCamera } from '../src/PerspectiveCamera.js';

// see e.g. math/Matrix4.js
const matrixEquals4 = function (a, b, tolerance) {
  tolerance = tolerance || 0.0001;
  if (a.elements.length !== b.elements.length) {
    return false;
  }

  for (let i = 0, il = a.elements.length; i < il; i++) {
    const delta = a.elements[i] - b.elements[i];
    if (delta > tolerance) {
      return false;
    }
  }

  return true;
};

describe('Cameras', () => {
  describe('PerspectiveCamera', () => {
    test('constructor', () => {
      const object = new PerspectiveCamera();
      expect(object).toBeDefined();
    });

    test('extends', () => {
      const object = new PerspectiveCamera();
      expect(object).toBeInstanceOf(Camera);
    });

    test('isPerspectiveCamera', () => {
      const object = new PerspectiveCamera();
      expect(object.isPerspectiveCamera).toBeTruthy();
    });

    test('type', () => {
      const object = new PerspectiveCamera();
      expect(object.type).toBe('PerspectiveCamera');
    });

    test('fov', () => {
      const object = new PerspectiveCamera();
      expect(object.fov).toBe(50);
    });

    test('zoom', () => {
      const object = new PerspectiveCamera();
      expect(object.zoom).toBe(1);
    });

    test('near', () => {
      const object = new PerspectiveCamera();
      expect(object.near).toBe(0.1);
    });

    test('far', () => {
      const object = new PerspectiveCamera();
      expect(object.far).toBe(2000);
    });

    test('focus', () => {
      const object = new PerspectiveCamera();
      expect(object.focus).toBe(10);
    });

    test('aspect', () => {
      const object = new PerspectiveCamera();
      expect(object.aspect).toBe(1);
    });

    test('view', () => {
      const object = new PerspectiveCamera();
      expect(object.view).toBeNull();
    });

    test('filmGauge', () => {
      const object = new PerspectiveCamera();
      expect(object.filmGauge).toBe(35);
    });

    test('filmOffset', () => {
      const object = new PerspectiveCamera();
      expect(object.filmOffset).toBe(0);
    });

    test('copy', () => {
      // prettier-ignore
      const near = 1, far = 3, aspect = 16 / 9, fov = 90;
      const object = new PerspectiveCamera(fov, aspect, near, far);
      const clonedObject = new PerspectiveCamera().copy(object, true);

      // will be different
      clonedObject.uuid = object.uuid;

      expect(clonedObject).not.toBe(object);
      expect(clonedObject).toStrictEqual(object);
    });

    test('setFocalLength', () => {
      // prettier-ignore
      const near = 1, far = 3, aspect = 16 / 9, fov = 90;
      const cam = new PerspectiveCamera(fov, aspect, near, far);

      cam.setFocalLength(50);
      expect(cam.projectionMatrix).toMatchObject({
        elements: [
          2.857142857142857, 0, 0, 0, 0, 5.079365079365079, 0, 0, 0, 0, -2, -1, 0, 0, -3, 0
        ]
      });

      cam.setFocalLength(40);
      expect(cam.projectionMatrix).toMatchObject({
        elements: [
          2.2857142857142856, 0, 0, 0, 0, 4.063492063492063, 0, 0, 0, 0, -2, -1, 0, 0, -3, 0
        ]
      });
    });

    test('getFocalLength', () => {
      // prettier-ignore
      const near = 1, far = 3, aspect = 16 / 9, fov = 60;
      const cam = new PerspectiveCamera(fov, aspect, near, far);

      expect(cam.getFocalLength()).toBeCloseTo(17.0498);
    });

    test('getEffectiveFOV', () => {
      // prettier-ignore
      const near = 1, far = 3, aspect = 16 / 9, fov = 60;
      const cam = new PerspectiveCamera(fov, aspect, near, far);

      expect(cam.getEffectiveFOV()).toBeCloseTo(59.9999);

      cam.setFocalLength(50);

      expect(cam.getEffectiveFOV()).toBeCloseTo(22.2753);
    });

    test('getFilmWidth', () => {
      // prettier-ignore
      const near = 1, far = 3, aspect = 16 / 9, fov = 90;
      const cam = new PerspectiveCamera(fov, aspect, near, far);

      expect(cam.getFilmWidth()).toBeCloseTo(35);
    });

    test('getFilmHeight', () => {
      // prettier-ignore
      const near = 1, far = 3, aspect = 16 / 9, fov = 90;
      const cam = new PerspectiveCamera(fov, aspect, near, far);

      expect(cam.getFilmHeight()).toBeCloseTo(19.6875);
    });

    test('setViewOffset', () => {
      // prettier-ignore
      const near = 1, far = 3, aspect = 16 / 9, fov = 90;
      const cam = new PerspectiveCamera(fov, aspect, near, far);

      const w = 1920;
      const h = 1080;
      const fullWidth = w * 3;
      const fullHeight = h * 2;

      /*
       * In grid of 3x2 displays:
       *
       *   +---+---+---+
       *   | A | B | C |
       *   +---+---+---+
       *   | D | E | F |
       *   +---+---+---+
       *
       */

      // set camera to render monitor A
      cam.setViewOffset(fullWidth, fullHeight, w * 0, h * 0, w, h);

      expect(cam.projectionMatrix).toMatchObject({
        elements: [
          1.1250000000000004, 0, 0, 0, 0, 2.0000000000000004, 0, 0, -2.0000000000000004, 1, -2, -1,
          0, 0, -3, 0
        ]
      });
    });

    test('clearViewOffset', () => {
      // prettier-ignore
      const near = 1, far = 3, aspect = 16 / 9, fov = 90;
      const cam = new PerspectiveCamera(fov, aspect, near, far);

      const w = 1920;
      const h = 1080;
      const fullWidth = w * 3;
      const fullHeight = h * 2;

      // set camera to render monitor A
      cam.setViewOffset(fullWidth, fullHeight, w * 0, h * 0, w, h);

      expect(cam.projectionMatrix).toMatchObject({
        elements: [
          1.1250000000000004, 0, 0, 0, 0, 2.0000000000000004, 0, 0, -2.0000000000000004, 1, -2, -1,
          0, 0, -3, 0
        ]
      });

      cam.clearViewOffset();

      expect(cam.projectionMatrix).toMatchObject({
        elements: [
          0.3750000000000001, 0, 0, 0, 0, 1.0000000000000002, 0, 0, 0, 0, -2, -1, 0, 0, -3, 0
        ]
      });
    });

    test('updateProjectionMatrix', () => {
      const cam = new PerspectiveCamera(75, 16 / 9, 0.1, 300.0);

      // updateProjectionMatrix is called in constructor
      const m = cam.projectionMatrix;

      // perspective projection is given my the 4x4 Matrix
      // 2n/r-l		0			l+r/r-l				 0
      //   0		2n/t-b	t+b/t-b				 0
      //   0			0		-(f+n/f-n)	-(2fn/f-n)
      //   0			0				-1					 0

      // this matrix was calculated by hand via glMatrix.perspective(75, 16 / 9, 0.1, 300.0, pMatrix)
      // to get a reference matrix from plain WebGL

      // prettier-ignore
      const reference = new Matrix4().set(
				0.7330642938613892, 0, 0, 0,
				0, 1.3032253980636597, 0, 0,
				0, 0, - 1.000666856765747, - 0.2000666856765747,
				0, 0, - 1, 0
			);

      // expect( reference.equals(m) );
      expect(matrixEquals4(reference, m, 0.000001)).toBeTruthy();
    });

    test('toJSON', () => {
      const object = new PerspectiveCamera();

      // will be different
      object.uuid = '934b9669-a7d5-4214-b723-dea6b4b3e068';

      expect(object).toMatchSnapshot();
    });

    test('clone', () => {
      // prettier-ignore
      const near = 1, far = 3, aspect = 16 / 9, fov = 90;
      const cam = new PerspectiveCamera(fov, aspect, near, far);

      const clonedCam = cam.clone();

      expect(cam.fov).toBe(clonedCam.fov);
      expect(cam.aspect).toBe(clonedCam.aspect);
      expect(cam.near).toBe(clonedCam.near);
      expect(cam.far).toBe(clonedCam.far);
      expect(cam.zoom).toBe(clonedCam.zoom);
      expect(cam.projectionMatrix.equals(clonedCam.projectionMatrix)).toBeTruthy();
    });

    test('from ObjectLoader', () => {
      // prettier-ignore
      const near = 1, far = 3, aspect = 16 / 9, fov = 90;
      const object = new PerspectiveCamera(fov, aspect, near, far);

      const json = object.toJSON();
      const loader = new ObjectLoader();
      const outputObject = loader.parse(json);

      expect(outputObject).toStrictEqual(object);
    });
  });
});
