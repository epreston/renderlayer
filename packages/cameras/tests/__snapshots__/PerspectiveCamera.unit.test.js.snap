// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cameras > PerspectiveCamera > toJSON 1`] = `
{
  "metadata": {
    "generator": "Object3D.toJSON",
    "type": "Object",
    "version": 4.5,
  },
  "object": {
    "aspect": 1,
    "far": 2000,
    "filmGauge": 35,
    "filmOffset": 0,
    "focus": 10,
    "fov": 50,
    "layers": 1,
    "matrix": [
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
    ],
    "near": 0.1,
    "type": "PerspectiveCamera",
    "up": [
      0,
      1,
      0,
    ],
    "uuid": "934b9669-a7d5-4214-b723-dea6b4b3e068",
    "zoom": 1,
  },
}
`;
